### PR TITLE
More type tweaks in Invoker

### PR DESCRIPTION
**Pull request by psuter**
Moving away from stringly-typed code, and from pervasive use of `Option`s.

### CODE REVIEW

**Comment by psuter** [comment]
( @perryibm may be interested too)

File: core/dispatcher/src/main/scala/whisk/core/container/Container.scala [hunk 1]
```diff
@@ -53,32 +53,28 @@ class Container(
 
     def details: String = {
         val name = containerName getOrElse "??"
-        val id = containerId getOrElse "??"
```
Comment (rabbah) [inline core/dispatcher/src/main/scala/whisk/core/container/Container.scala]: we're guaranteed to have a container id now?

File: core/dispatcher/src/main/scala/whisk/core/container/ContainerPool.scala [hunk 2]
```diff
@@ -555,12 +557,11 @@ class ContainerPool(
     // We send the payload here but eventually must also handle morphing a pre-allocated container into the right state.
     private def initWhiskContainer(action: WhiskAction, con: WhiskContainer)(implicit transid: TransactionId): FinalContainerResult = {
         con.boundParams = action.parameters.toJsObject
```
Comment (rabbah) [inline core/dispatcher/src/main/scala/whisk/core/container/ContainerPool.scala]: so invariant is that `con` is a valid container (hence has a valid id)?

File: core/dispatcher/src/main/scala/whisk/core/container/ContainerUtils.scala [hunk 3]
```diff
@@ -73,51 +74,53 @@ trait ContainerUtils extends Logging {
...
-    def killContainer(name: String)(implicit transid: TransactionId): DockerOutput = killContainer(Some(name))
+        val cmd = Seq("run") ++ makeEnvVars(env) ++ consulServiceIgnore ++ nameOption ++ cpuArg ++ memoryArg ++
+            capabilityArg ++ fileHandleLimit ++ processLimit ++ securityOpts ++ containerNetwork ++ Seq("-d", image) ++ args
 
-    def killContainer(container: ContainerName)(implicit transid: TransactionId): DockerOutput = {
-        container flatMap { name => runDockerCmd("kill", name) }
+        runDockerCmd(cmd: _*).toOption.map { result =>
+            ContainerHash.fromString(result)
+        } getOrElse {
+            throw new Exception("Container hash or name expected in `makeContainer`.")
```
Comment (rabbah) [inline core/dispatcher/src/main/scala/whisk/core/container/ContainerUtils.scala]: i see it.

File: core/dispatcher/src/main/scala/whisk/core/container/package.scala [hunk 4]
```diff
@@ -77,8 +77,42 @@ package object container {
     case object Busy extends FinalContainerResult
     case class Error(string: String) extends FinalContainerResult
 
-    type ContainerName = Option[String]
```
Comment (rabbah) [inline core/dispatcher/src/main/scala/whisk/core/container/package.scala]: ah at long last, proper types.

**Comment by rabbah** [comment]
reviewed - pg pending?

File: core/dispatcher/src/main/scala/whisk/core/container/Container.scala [hunk 5]
```diff
@@ -53,32 +53,28 @@ class Container(
...
         // docker ps contains only a prefix of the id
-        containerId map { _.take(8) } getOrElse "anon"
+        containerId.id.take(8)
     }
 
     /**
      * Gets logs for container.
      */
     def getLogs()(implicit transid: TransactionId): String = {
-        getContainerLogs(containerId) getOrElse ""
```
Comment (psuter) [inline core/dispatcher/src/main/scala/whisk/core/container/Container.scala]: It looks like this access to logs is used only in tests (checking that logs contain specific `String`s), where always getting a (possibly empty) `String` is a reasonable interface.

**Comment by psuter** [comment]
PG1 363 is all blue and shiny.

**Comment by rabbah** [comment]
Rebased and merged as https://github.com/openwhisk/openwhisk/compare/7b2425fc5deb...a850e88218ce.

**Pull request closed without merging**